Site Logo: Add option to set site icon from Site Logo block

stacimc opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->
Related to: #30406 

## Description
<!-- Please describe what you have changed or added -->
Adds an option in the Inspector Controls of the Site Logo block to sync the site logo image to the Site Icon. When the toggle is switched on, any updates persisted to the Site Logo image are also made to the Icon.

When you set a new Site Logo for the first time (ie add a new Logo block and upload an image when it was previously undefined), _if there is no existing site icon_ this toggle will be automatically turned on and the logo will be synced to the icon. When the user saves the post, they Save prompt makes it clear that they are about to persist changes to the logo and/or icon.

Any unsaved changes to the Logo & Icon are discarded when you remove the last Site Logo block from a post (see testing instructions).

## Notes 
* The help text for the toggle ought to link you to a settings page where you can upload a different image or reset your site icon. Right now it links to the Customizer's Site Identity section.
    *  On blocks themes, the Customizer link is removed from wp-admin, but it can still be linked to directly. There is an issue open for adding a new flow for updating site icons through the site editor (#29126). When this is resolved, this link ought to be updated.
* The link to the Customizer currently opens in a new tab. This is to prevent the blocking alert to Save Changes/losing changes.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

### Test cases for an authorized user:

#### Test initial value of the toggle:
- [ ] Make sure you have no Site Logo or Site Icon set on your site. Open a post and add a Site Logo block. Upload an image and verify in the Inspector controls that the toggle is automatically turned on. Save the post and verify that you are asked if you want to save changes to both the logo and the icon. Verify the changes save correctly.
- [ ] Remove your Site Logo but **make sure you have a Site Icon set**. Open a post and insert a Site Logo block. Upload an image and verify that the toggle is not automatically turned on. Save the post and verify that you are asked if you want to save changes to the logo but **NOT** the icon.

#### Test updating the Site Logo:
- [ ] Replace the Site Logo with a different image and save the post. You should again be prompted to save both the icon and the logo. Verify the changes persist.
- [ ] Remove the Site Logo with the `Reset` option from the `Replace` menu in the block toolbar. You should again be prompted to save both the icon and the logo. Verify changes persist after saving (icon and logo should both be removed).
- [ ] Replace the Site Logo with a new image. Toggle the option to sync off. Save the post. You should only be prompted to save the Logo. Verify that the new Logo is added but the icon is still empty.

#### Test persistence of Toggle state:
- [ ] Toggle the sync option on and upload a new image. Save all entities. Then toggle the option off. You should be able to save the post. When you refresh, the toggle should still be turned off even though both logo and icon are the same image.

#### Test staying synced with changes via the Customizer:
- [ ] Toggle the sync option on and save with a new image. Verify both icon and logo are updated. Then go to the customizer and update the site icon through the Site Identity section. Open the post in the block editor again and verify that the toggle is now OFF.
- [ ] Toggle the sync option and upload a new image. Click 'save' and verify you see prompts to save both icon and logo, but do not actually save. Go to the customizer and upload/publish a new icon. Return to the block editor tab and now hit Save, without refreshing first. Since the toggle is still enabled, the changes to the icon here should override what you saved in the customizer.

#### Test discarding unsaved changes:
- [ ] Insert a Site Logo block into a new post. Update the image and make sure the toggle is on. Verify you see the option to save both Logo & Icon in the pre-save menu, but do not save yet.
- [ ] Insert at least one more Site Logo block. Note that it displays with the updated image.
- [ ] Hover over the 'Default' and 'Rounded' styles in the Controls, and then move the cursor away so the preview renders and then disappears. _This is to test that unmounting the Site Logo block in the preview does not cause unsaved changes to be discarded._ Verify that the Logos in your post still show your unsaved change.
- [ ] Open the large block inserter panel from the `+` in the toolbar. Scroll down to Site Logo and hover over it to make the preview appear, then move the cursor away so it disappears. _This is to test that unmounting the Site Logo block in this preview also does not cause unsaved changes to be discarded._ Again, the Logos in your post should be unaffected.
- [ ] Remove the first Logo block. Because there's still a Logo block on the page, unsaved changes should not be discarded yet.
- [ ] Remove the second/final Logo block. Your unsaved changes should be discarded: you should not see `Logo` or `Icon` in the pre-save menu, and if you insert a new Site Logo block it should display the published logo. Save the post and verify that the changes were not made.
- [ ] Repeat the above tests in the site editor; make sure you include multiple Site Logo blocks in different templates. (For example, add one to the Header and one to the Footer; make an unsaved change; remove one block and verify the change is not discarded; remove the second block and verify that it is).

### Tests as an unauthorized user (for example with the Author role):
- [ ] Insert a Site Logo block in a new post. Verify that you don't see the toggle in the Inspector Controls.

## Screenshots <!-- if applicable -->
Controls (updated to show the new copy):
<img width="283" alt="Screen Shot 2021-12-07 at 1 44 33 PM" src="https://user-images.githubusercontent.com/63313398/145110976-296b9551-3d0b-485c-86c6-66423d45e8ca.png">

Save prompt:
<img width="274" alt="Screen Shot 2021-11-01 at 11 47 20 AM" src="https://user-images.githubusercontent.com/63313398/139724728-0b666199-6acd-4478-a9f1-c3144f87140f.png">



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

stacimc committed:
Add toggle for syncing site icon to site logo

stacimc committed:
Sync icon when first uploading logo, if no icon exists

noisysocks commented:
Hi @stacimc, thanks so much for digging into this. Are you still working on it?

stacimc commented:
@noisysocks I'm just picking this back up.

At the moment it matches designs proposed in the issue, except that it does not link to an external page to edit the Site Icon settings. I set this aside because I was expecting #29126 to add those settings somewhere in the Site editor or admin, where I could link to them.

At the moment I think the best I can do is link to the Customizer, since unfortunately I don't think it's possible to link to the Site Identity section specifically.

stacimc committed:
Do not clear the site icon when toggling off syncing, instead restore to original icon

stacimc committed:
Link to Customizer from Site Logo syncing help text

stacimc committed:
Sync site icon on uploading new image or resetting image

stacimc committed:
Simplify logic

stacimc committed:
Automatically sync icon if none exists and adding logo for the first time

ramonjd commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -80,6 +80,25 @@ function register_block_core_site_logo_setting() {
+		'site_icon',
+		array(
+			'show_in_rest' => array(
+				'name' => 'site_icon',
```
Could this be `'show_in_rest' => true,` since we're not renaming the `$option_name` param?

ramonjd replied:
Also I think we need to add the translation for `__( 'Site icon' )` in the `TRANSLATED_SITE_PROTPERTIES` [sic] list here: https://github.com/WordPress/gutenberg/blob/9df736c678b2e8ef00c355e6cc9fb8092fe25484/packages/editor/src/components/entities-saved-states/index.js#L25

Otherwise we see the `site_icon` property name.
<img width="277" alt="Screen Shot 2021-11-01 at 9 40 20 am" src="https://user-images.githubusercontent.com/6458278/139603862-b30504b0-cefa-4446-84e7-772c0353dd38.png">

ramonjd commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -250,6 +264,14 @@ const SiteLogo = ( {
+			"Site Icons are what you see in browser tabs, bookmark bars, and within the WordPress mobile apps. If you don't have one, you can set your logo to also be your icon. If you do have a custom site icon, you can upload that from the Site Icon settings in the <a>Customizer</a>!"
+		),
+		// eslint-disable-next-line jsx-a11y/anchor-has-content
+		{ a: <a href={ siteUrl + '/wp-admin/customize.php' } target="blank" /> }
```
> At the moment I think the best I can do is link to the Customizer, since unfortunately I don't think it's possible to link to the Site Identity section specifically.

The URL could also be `/wp-admin/customize.php?autofocus[section]=title_tagline`, which opens up the Site Identity panel straight away.

stacimc replied:
Brilliant, this makes it _much_ nicer!

ramonjd commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -250,6 +264,14 @@ const SiteLogo = ( {
+	const syncSiteIconHelpText = createInterpolateElement(
+		__(
+			"Site Icons are what you see in browser tabs, bookmark bars, and within the WordPress mobile apps. If you don't have one, you can set your logo to also be your icon. If you do have a custom site icon, you can upload that from the Site Icon settings in the <a>Customizer</a>!"
+		),
```
I have personal preference for less copy, especially seeing as the Site Logo block description covers a lot already, so this suggestion is 100% ignorable 😄 

`Toggle to set your logo to also be your icon. You can upload a custom site icon in the <a>Customizer</a>.`

stacimc replied:
Hm I was concerned with making sure that the distinction between icon and logo was clear here. You make a good point about the Site Logo block description -- I hadn't noticed it addresses the site icon in that text as well:

`
Display a graphic to represent this site. Update the block, and the changes apply everywhere it’s used. This is different than the site icon, which is the smaller image visible in your dashboard, browser tabs, etc used to help others recognize this site.
`

Knowing that it does feel a bit redundant, but I still think it's important to make sure the use of the toggle is clear from its own helptext, and I imagine not everyone will read through the block description 🤔

ramonjd commented:
Thanks @stacimc ! So far, setting/resetting the logo and icon in the post mostly works for me in non-block and block themes (I tested Twenty Twenty One and TT1 Blocks).

The changes are synched to the Customizer and on the frontend.

I noticed an issue with new posts whereby I couldn't publish a new post if I deselect the site logo/icon options on the pre-publish panel. 

https://user-images.githubusercontent.com/6458278/139604586-ed6c0510-df86-4379-86d5-cecc8f66b294.mp4

I'm not sure why it's occurring yet, but the error is `this.props.setEntitiesSavedStatesCallback is not a function`
It looks like the prop `setEntitiesSavedStatesCallback` is not passed to the [post publish button](https://github.com/WordPress/gutenberg/blob/a6908dcc4cde747ee2879317597f73179bf6e4d1/packages/editor/src/components/post-publish-button/index.js#L56) from the [post publish panel](https://github.com/WordPress/gutenberg/blob/a6908dcc4cde747ee2879317597f73179bf6e4d1/packages/editor/src/components/post-publish-panel/index.js#L93).

Note, I can replicate this on trunk as well so it's not related to this PR I think. I just thought I'd flag it here first while it's fresh 😄 

Created an issue here: https://github.com/WordPress/gutenberg/issues/36096

stacimc committed:
Add translated string for site icon property

stacimc committed:
Link directly to Site Icon settings in Customizer

stacimc committed:
Adjust copy

kevin940726 commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -250,6 +264,24 @@ const SiteLogo = ( {
+						siteUrl +
+						'/wp-admin/customize.php?autofocus[section]=title_tagline'
+					}
+					target="blank"
```
Shouldn't it be `target="_blank"`?

stacimc replied:
Yep, absolutely. Fixed.

stacimc committed:
Correct opening customizer in new tab

stacimc commented:
> I noticed an issue with new posts whereby I couldn't publish a new post if I deselect the site logo/icon options on the pre-publish panel.
<...>
Note, I can replicate this on trunk as well so it's not related to this PR I think. I just thought I'd flag it here first while it's fresh 😄

@ramonjd This is a weird one, nice spot! I played around with this a bit and I can replicate it on trunk as well, and for that matter with other non-post entities like template parts, so I'm also confident this PR isn't related. Thanks for opening the issue, I'll add a comment there as well 👍

ramonjd commented:
Swinging back to test this one further, I'm seeing a few things I didn't notice before.

The following occurs when I've **already set the logo as my site icon in the block editor** and saved the post. 

Things appear as expected, but if I try to toggle the site icon off, without making any other changes, the editor state isn't dirty (and therefore not updatable) when toggling **Use as site icon**

Toggling other options triggers a dirty state.

<img width="150" src="https://user-images.githubusercontent.com/6458278/139956457-81b6f070-7314-4b53-9aa2-0acf47194efc.gif" />

Also, toggling the checkbox in the editor no longer appears to have any effect. I don't see the pre-publish save bar to confirm Site changes (not sure if I should), and the site icon remains on the site and in the Customizer.

I'm not if its related, but it looks like the [dirtyRecords selector](https://github.com/WordPress/gutenberg/blob/0fb9e8cb137a2bae2401a313538a5f37a84ee1a9/packages/editor/src/components/entities-saved-states/index.js#L34) isn't being fired for some reason.

Can you reproduce this as well?

stacimc commented:
> Swinging back to test this one further, I'm seeing a few things I didn't notice before.
> 
> The following occurs when I've **already set the logo as my site icon in the block editor** and saved the post.
> 
> Things appear as expected, but if I try to toggle the site icon off, without making any other changes, the editor state isn't dirty (and therefore not updatable) when toggling **Use as site icon**

If I'm understanding you correctly, what you're doing is:

1. Sync the logo and icon using the toggle
2. Save/publish the post
3. Toggle the option _off_, but make no other changes
4. Here you expect to be able to Update the post, but it appears that nothing has happened and you do not have the option to Update

I think this is working as I intended, but I can see why it may be confusing and am very open to suggestions for different behavior.

When you turn off the `Use as site logo` option, you indicate that any updates to the Site Logo shouldn't be synced to the icon, but it can't restore the icon to some previous value it may have had before syncing. The exception is if you have _unsaved_ changes to the icon. So for example if I insert a Site Logo block, upload an image, and enable the toggle, as long as I turn the toggle off before saving, no changes will be made to the icon. But if changes have already been persisted, we'd need to keep track of previous values in order to restore them, which opens up a can of worms.

At the moment that option doesn't get persisted in an attribute or anything; when you load the block it is turned on if the logo/icon match.

stacimc commented:
Another thing we could do is clear the site icon when the `Use logo as site icon` option is toggled off. I actually had it doing this in an an initial pass but worried that might be unexpected behavior. What do you think @ramonjd?

ramonjd commented:
> When you turn off the Use as site logo option, you indicate that any updates to the Site Logo shouldn't be synced to the icon, but it can't restore the icon to some previous value it may have had before syncing. 

Thanks for clarifying this point. This makes total sense when you say it like that 😄 So it's more of a "Use once, then keep in sync with site icon" sort of thing? :) 

Armed with that revelation, I retested and can confirm that the switch toggles _synching future changes_ to the site logo image. 👍 

I'm not sure how to square it, but if I think of it as "Keep changes to the logo block  in sync with the site icon", I still kinda expect the editor to allow me to save the state when I'm toggling off "Use as site icon". Even if it doesn't change/clear/affect the current value of the site icon.

I'm not saying this is the right expectation to have of course! :)

> But if changes have already been persisted, we'd need to keep track of previous values in order to restore them, which opens up a can of worms

Ah good point. I clearly didn't think that far ahead. I think assumed the behaviour that [you propose](https://github.com/WordPress/gutenberg/pull/35892#issuecomment-958329884) in this comment, but I agree that it might cause an unintentional clearing of the site icon.

It's a difficult one. 

Maybe the toggle control label changes once a site logo has been selected, or changes. So if the current site logo is not the site icon, it'd say "Use as site icon", then after toggling to "on" it would say "Keep in sync". No idea if that's possible or even a good idea, or if it would make things clearer 🤷

stacimc committed:
Persist toggle state in an attribute

stacimc committed:
Turn the toggle off if the logo and icon fall out of sync

If a user syncs the logo and icon via the toggle and saves changes, but later
changes the site icon in the Customizer, we want the toggle to be reset to the
`off` position when they next edit the block.

stacimc commented:
I updated this to [persist the state of the toggle in a block attribute](https://github.com/WordPress/gutenberg/pull/35892/commits/3f4a0ea0739118bcc2d820eee6d68b66bdafedf6). Here's how it affects @ramonjd's scenario:

1. Insert a Site Logo block, upload a picture, and turn the toggle on. Save all entities.
2. Toggle the `Use as site icon` switch _off_, but make no other changes.

New behavior:
* Toggling the switch dirties the post, so you are able to Update.
* If you do update and return to the post at a later time, the toggle will still be in the `off` position. If you change the Site Logo, it will not affect the Icon unless you deliberately toggle the option back on.
* Toggling the option off in this scenario still won't affect or clear the actual Site Icon, because it has already been saved. It _will_ clear any _unsaved_ changes to the Icon.

The problem is this makes staying synced with changes made to the Icon in the Customizer more confusing. Scenario:
* Update the Site Logo in the block editor and sync to the icon via the toggle. Save all entities.
* Open the customizer and update the Site Icon.
* Open the block editor again.

Changing the icon elsewhere feels like a deliberate action to un-sync the logo and icon, so I would expect the toggle to be off when I revisit the block editor, even though I last saved it in the `on` position. This is especially important because otherwise the page will be dirty as soon as you load it. I added some logic to detect this and [reset the toggle to false](https://github.com/WordPress/gutenberg/pull/35892/commits/3f4a0ea0739118bcc2d820eee6d68b66bdafedf6) if the icon has been changed.

I think I'm happy with that balance. @mtias can you weigh in here -- does this align with your idea?

ramonjd commented:
Thanks for working on this @stacimc !

After retesting here's what I'm seeing now:

https://user-images.githubusercontent.com/6458278/140204703-28718024-d6ab-4c3f-9570-5f0f36aad871.mp4

✅ I inserted a new site logo and synched to my site icon. After saving, the site logo and icon match.
✅ In the post I toggled "Use as site icon" to off. The post allowed me to update and save my settings.
✅ I then replaced the site logo and saved. The site logo updated but the icon didn't change. 

I also went through the testing scenarios again and things are working as described. 🎉

jameskoster commented:
> Changing the icon elsewhere feels like a deliberate action to un-sync the logo and icon, so I would expect the toggle to be off when I revisit the block editor, even though I last saved it in the on position.

Agreed with this.

One other flow to consider:

1. Insert and set Site Logo
2. Toggle "Use as site icon" on
3. Save
4. Unset site logo and delete block
5. Save

In this case I would expect the Site Icon to remain set, but wanted to double-check that is what happens?

ramonjd commented:
> In this case I would expect the Site Icon to remain set, but wanted to double-check that is what happens?

I just retested and can confirm that this is true if you delete the block in the editor and remove the site logo from the Customizer. 

If you reset the site logo from the block toolbar and save the logo + icon, then both the logo and icon are unset.

<img width="485" alt="Screen Shot 2021-11-08 at 8 30 33 am" src="https://user-images.githubusercontent.com/6458278/140662563-b6e52d98-e5e5-43f3-a506-1fe1f280cbe5.png">

I noticed something else that caused me to scratch by beard: removing the site logo in the editor has no effect on the application state unless you refresh the editor page. 

1. On a new post, insert a site logo and save.
2. In a new tab, check that the logo is set on the frontend, and in the Customizer.
3. In the editor tab, remove the block and update (don't refresh the page). The site logo won't be cleared, so also unset the logo in the Customizer in a new tab, and update. 
4. Refresh the frontend tab and check that the site logo is removed.
5. Return to the post tab and insert the site logo block again. I'm seeing that my last selection is pre-loaded.
6. Save the post. Once again, don't refresh the editor tab.
7. Refresh the frontend page to see that the logo isn't applied.
8. Now refresh the editor page. The site logo block is empty. Add an image and save. 
8. Only now is the logo applied on the frontend.

I was just wondering if that's expected, or whether we can force an update on the logo even if the editor page isn't refreshed.

ramonjd commented:
> removing the site logo in the editor has no effect on the application state unless you refresh the editor page.

I _believe_ it's because the media record `mediaItemData` is stored in state. To trigger an entityRecord update, the id of the new logo has to be different from the previous logo.

So, if we want to set the **same** logo in the editor immediately after having deleted it in the Customizer, we have to clear the entityRecord first (?).

stacimc commented:
> removing the site logo in the editor has no effect on the application state unless you refresh the editor page.

While playing around with this, I think I found a simpler way to reproduce/a slightly different problem:

1. Start with a site that already has a Site Logo set.
2. Insert a Site Logo block into the block editor and update the image to something new.
3. Immediately remove the block without saving your changes.
4. Notice that you still see the option to 'Update' the post, and if you click through you'll see the option to save the Site Logo changes.
5. Insert a new Site Logo block and see that it shows your updated image.

When I remove the block, I think I'd expect any unsaved changes to the logo to be discarded. Although even that can be a bit complicated if you have multiple Site Logo blocks on a page -- I imagine changes should be discarded only if all blocks are removed 😓 I'll see what I can do here.

For the example you gave @ramonjd with editing the Logo concurrently in the customizer: I'm not sure that I'd expect changes made to the Logo in another tab to be reflected until I've refreshed the page. I'm not deeply familiar with the way we detect changes, but I think that's a much trickier can of worms.

ramonjd commented:
> I'm not sure that I'd expect changes made to the Logo in another tab to be reflected until I've refreshed the page. I'm not deeply familiar with the way we detect changes, but I think that's a much trickier can of worms.

Thanks for looking at it. I wasn't sure either. It's a bit of an outlier, and most probably a challenge - if it even is one - to think about later. 🙇

stacimc committed:
Discard unsaved changes to logo and icon when removing last Logo block

stacimc committed:
Prevent uncontrolled input when sync attribute undefined

stacimc commented:
This has now been rebased on trunk. I've also updated it to discard any unsaved changes to the logo and icon if you insert the block, make changes, but then remove it before saving. Testing instructions added to the description.

apeatling reviewed (changes requested):
I've had a look at the status of this and site logo management in general:

* Adding/editing a site logo to general settings will be done for 6.0, not 5.9: https://core.trac.wordpress.org/ticket/54370. 
* The customizer will not be accessible in 5.9 with a block theme, but the direct link to edit the site logo will.
* This flow makes sense and is logical to me. It has significant value regardless of what happens with the site logo UI in 6.0.
* It was noted by Matías as an important flow in https://github.com/WordPress/gutenberg/issues/30406

Testing this -- it's working well for me, and I think we should move ahead with it. This is assuming that in a follow up release that the customizer link would be replaced with the link to the new UI for managing the site logo (probably in general settings).

I think the copy for the explanation text can be tightened up. Something more like: 

> Your site icon is what visitors will see in browser tabs, bookmark bars, and in the WordPress mobile app. 
>
> To use a custom icon that is different from your site logo, use the [site icon settings].

stacimc committed:
Tighten up copy

stacimc commented:
Thanks @apeatling. That recap looks correct to me -- in particular, it looks like at the moment there is _no_ way in 5.9 to edit the Site Icon from blocks themes without manually navigating to the direct link, which is (with good reason) not linked anywhere. This PR at least provides a link to get to that settings page, without having to expose the customizer in the admin. Agreed that we would definitely follow up to replace that link as soon as a more stable settings page is created.

> I think the copy for the explanation text can be tightened up. 

Thanks for this -- now that the label of the toggle has been updated from `Sync site icon` to the much easier-to-understand `Use as site icon`, I think you're right and we can omit that second sentence explaining the toggle. Done!

apeatling reviewed (approved):
Copy changes look :+1:. I think this is looking good, but let's get some more thoughts on if this should be a 5.9 thing or not.

jameskoster commented:
> but let's get some more thoughts on if this should be a 5.9 thing or not.

Doesn't it have to be? Otherwise there's no way to set the site icon when a block theme is active 🤔

stacimc committed:
Force syncing on initial selection

stacimc commented:
Rebased since it's been awhile, and pushed a small change to make sure the toggle is switched on when adding a Logo for the first time if there is no Icon present.

> > but let's get some more thoughts on if this should be a 5.9 thing or not.
>
> Doesn't it have to be? Otherwise there's no way to set the site icon when a block theme is active

This was my concern 😞 I would love to get some additional testers on this as the syncing logic is fairly complex.

paaljoachim commented:
Testing.

Beta 3 WP 5.9
Local Dev site built in Desktop Server.
Twenty Twenty Two.
Site Editing screen.
Chrome
OSX

> Adds an option in the Inspector Controls of the Site Logo block to sync the site logo image to the Site Icon. When the toggle is switched on, any updates persisted to the Site Logo image are also made to the Icon.

### Before:

![before](https://user-images.githubusercontent.com/5323259/146614911-628aa3e6-fe10-4e0c-897f-48c4737268a1.jpg)

### After:

![After](https://user-images.githubusercontent.com/5323259/146616069-11420af9-11af-4536-9f51-ff495bcfa99a.jpg)


> When you set a new Site Logo for the first time (ie add a new Logo block and upload an image when it was previously undefined), if there is no existing site icon this toggle will be automatically turned on and the logo will be synced to the icon. When the user saves the post, they Save prompt makes it clear that they are about to persist changes to the logo and/or icon.

Sounds good!


I have not read through everything, but the bottom line for me is that it is important to have some way to change the site icon. Having it included with the Site Logo block is helpful. As it seems to be the only way to change the site icon that will be included for FSE themes in 5.9.

andrewserong commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -349,14 +403,57 @@ export default function LogoEdit( {
 			},
 			isRequestingMediaItem: _isRequestingMediaItem,
+			siteIconId: _siteIconId,
+			logoBlockCount: getGlobalBlockCount( 'core/site-logo' ),
```
In testing, it seems like this value might be being cached on the first render, possibly due to the empty dependency array below. Sometimes when logging out this value when removing the second Site Logo block on the page, it'd return `1` and removing the first logo on the page (after removing the second logo), it would return `2`.

I'm not sure if this is related to the inconsistent discarding of site logo start that I noticed. I wonder if it'd work to move this particular value to its own `useSelect`?

stacimc replied:
Good catch -- I must have only been testing removing blocks in the order they were inserted! This bug makes sense because I was counting the `logoBlockCount` only on the first render.

I've updated the PR to calculate the blockCount _at the time of removal_.

andrewserong commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -349,14 +403,57 @@ export default function LogoEdit( {
+
+			// Only discard unsaved changes if we are removing the last Site Logo block
+			// on the page.
+			if ( logoBlockCount === 1 ) {
```
I _think_ I encountered a couple of times where this fires outside of the last block being removed.

This first was after removing the first Site Logo block when that block had unsaved changes. I actually didn't really mind seeing the change there — I think removing the Site Logo block kind of takes precedence over preserving the changes we might have done.

![Kapture 2021-12-20 at 12 24 58](https://user-images.githubusercontent.com/14988353/146699645-0faf6ff1-e056-4e85-b9f8-d840295eb2e7.gif)

The second was after changing the Site Logo image but before saving, if I open the block inserter and then start typing to search for the Site Logo block, the changes are lost:

![Kapture 2021-12-20 at 12 20 48](https://user-images.githubusercontent.com/14988353/146699766-6d6b6d94-95e5-4e0e-8a3a-7a3d058dc599.gif)

Could there be another reason why the component gets unmounted at that stage? It's unfortunate that we don't have an easy "onBlockRemoved" hook, and I couldn't quite think of another workaround 😞

stacimc replied:
The first issue you described should [now be resolved ](https://github.com/WordPress/gutenberg/pull/35892#discussion_r772612901). I wasn't able to reproduce the second one -- can you verify if it's still happening?

I'm pretty stumped as to what could have caused this. I had to add in checks to make sure changes weren't discarded when unmounting the block from the Styles Preview or the Block Inserter preview -- but those scenarios are still working for me, and in your video you don't seem to open either preview 🤔

glendaviesnz reviewed (approved):
Tests well for me. Also seems like an important change for 5.9 to me. 

I think the failing navigation e2e has been fixed so a rebase should fix that.

andrewserong reviewed (commented):
Thanks so much for the detailed testing steps @stacimc! This feels nearly there to me. I've added a couple of comments about inconsistency with discarding the unsaved state that I noticed in testing, but other than that, it worked really nicely, and the syncing behaviour feels solid to me. I noticed that there isn't an explicit way to "clear" the site icon without resetting the site logo, but since we've got the link to the customizer, for folks who want the clear button, they can go there in the shorter-term 👍

I also noticed it's possible to set one Site Logo block to sync and another one on the same page to _not_ sync:

![Kapture 2021-12-20 at 12 23 26](https://user-images.githubusercontent.com/14988353/146700768-7c1d6fc8-13c8-4aca-9312-893c6ba54a10.gif)

At first I thought this might be an issue, but it actually works fine in testing! The synced one syncs changes to the site icon, and the other one appears not to, but I couldn't find any actual problem that introduces 👍

With this PR it's also possible to set a non-square site icon, if the site logo image isn't square. So the rendering of the site icon in the top left with non-square logos goes slightly outside its container:

![image](https://user-images.githubusercontent.com/14988353/146701194-6618eec5-b769-480a-9c86-e3ec23ef77fe.png)

In the rendered markup of my site, the `32x32` icon is a scaled square ratio image, but the full size icon / apple touch icon links to the full uncropped image:

![image](https://user-images.githubusercontent.com/14988353/146701376-90c634a2-6c2e-4060-a5aa-67076d257cc1.png)

I'm not sure how much of an issue that is? Folks can still go to the customizer to crop the site icon if they need to.

I think the only real blocker from my perspective, then, is unintentional discarding of unsaved changes when typing in the block inserter. That the unsaved changes were discarded when deleting one of the site logo blocks on a page didn't worry me (if I'm deleting something after creating an unsaved change, I don't think that's too big of a deal).

Please let me know if you need more context on the comments I left, and happy to do more testing!

jameskoster commented:
The handling of non-square logos doesn't seem all that great to me:

https://user-images.githubusercontent.com/846565/146760649-d0170371-b2ea-42f9-9adf-cb4acfbc526c.mp4

Then again, I'm not sure how we can really encourage / facilitate cropping in this particular flow, so on balance perhaps this is ok for now.

I would say that we should probably apply an `overflow` rule to the W button though, to stop that strange on-hover behaviour.

stacimc committed:
Get up-to-date logoBlockCount when removing block

Previously we were getting the logoBlockCount on block creation, but
this value could become out of date if more Site Logo blocks were added
afterward.

Now we count the number of blocks at the time of block removal, so that it
is guaranteed up-to-date, and we ensure that changes are only discarded
if there are no other Site Logo blocks on the page.

stacimc commented:
Thanks for all the testing! Rebased to hopefully pick up e2e test fixes, and I've updated the logic to discard unsaved changes when removing the final Site Logo block.

> With this PR it's also possible to set a non-square site icon, if the site logo image isn't square

> I would say that we should probably apply an overflow rule to the W button though, to stop that strange on-hover behaviour.

I _think_ this is consistent with the behavior currently on trunk. I can test setting a non-square Site Logo by going through the Customizer and clicking `Skip cropping` when I select a non-square image from my media library.

The behavior on hover is also in trunk, even when I select a Site Icon through the Customizer settings page and crop it appropriately. It looks like the animation is fairly new from #33132, and the outline comes from #29888. I do think that the outline looks unusual when a non-square icon is hovered:
<img width="60" alt="Screen Shot 2021-12-20 at 2 40 08 PM" src="https://user-images.githubusercontent.com/63313398/146842167-bcd9e644-b1c0-42e6-90b7-53cee6bf4fe4.png">

stacimc committed:
Add shouldSyncIcon attr to core blocks documentation

stacimc commented:
Not sure what's up with all the failing tests after rebase, taking another look now 👀 

Edit: Updated the docs! The other tests _should_ be good when we're able to re-run 🤞

andrewserong reviewed (approved):
Great work @stacimc! I've re-tested this, and I can't reproduce any of the unsaved changes issues that I ran into previously. I believe they've all been fixed up via moving `getGlobalBlockCount` to its own `useSelect` (and calling it at the last minute when needed), and shifting the check for `logoBlockCount` to be against `0` instead of `1`. This is testing really well 🙂

It'd be good to fix up the styling on the W button, but given the progress already in this PR, I'd be in favour of merging in this PR as-is and us then looking into the button styling in a follow-up, considering it's already a bug on `trunk`.

LGTM! 🎉

jameskoster commented:
> I do think that the outline looks unusual when a non-square icon is hovered

Yeah that was my main concern. I see that the "outline" is a pseudo element, so a simple `overflow` style will not fix this. We can probably address it in a follow up.

stacimc commented:
> > I do think that the outline looks unusual when a non-square icon is hovered
> 
> Yeah that was my main concern. I see that the "outline" is a pseudo element, so a simple `overflow` style will not fix this. We can probably address it in a follow up.

@jameskoster Agreed! Issue created: #37564

youknowriad commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -349,14 +401,59 @@ export default function LogoEdit( {
 	const { editEntityRecord } = useDispatch( coreStore );
-	const setLogo = ( newValue ) =>
+
+	useEffect( () => {
```
Hi there!

The logic here doesn't make sense, it's creating bugs like #37760

Basically the logic here assumes than when we "unmount" the site logo block, we're removing it when it's not always the case, unmounting can happen in several ways:

 - Switching to code editor in post editor
 - Changing the current view (template/template part) in the site editor.

While I'm not sure about the rest of this PR, this effect should be removed.

ramonjd replied:
@stacimc I removed the hook and went through the testing instructions above under **"Test discarding unsaved changes:"**

I couldn't find any weird side effects, that is, the editor recognised and kept my updates when viewing previews and when switching between code/visual modes. 

In other words, things still work for me.

Is there something I'm not testing right?

stacimc replied:
The logic here doesn't assume that unmounting the block means we're removing it. It runs every time the block is unmounted, but it then [checks the global block count](https://github.com/WordPress/gutenberg/pull/35892/files#diff-0d5debfeb0c6fe28a3979c43ecb9029bae1617d404b1c92f506ba754333f9647R432) to determine whether the Site Logo blocks have actually been removed; otherwise it should do nothing.

I think the problem is in the Site Editor, that `getGlobalBlockCount` isn't behaving the way I believed it would. Although all my own test scenarios still work for me, I'm able to reproduce the #37760 bug by:
* Add a Site Logo block and make a change in the Header part
* Add a different block to a different template part which does _not_ contain a Site Logo block
* Hit Undo and see that the 'Redo' button is disabled

When the Site logo block unmounts, `getGlobalBlockCount` returns 0 this time.

stacimc replied:
I'll put up a PR to remove the logic for now and then start looking into a better way of doing this. I do think it would be nice to discard those unsaved changes when the block is removed, if I can get it working.

stacimc replied:
#37895 fixes the regression, and #37900 has been opened to investigate adding the functionality back in safely.

youknowriad replied:
Thanks @stacimc `getGlobalBlockCount` is not reliable because the `core/block-editor` store is more like an internal store to an instance of the block editor and don't say anything about the presence of a block in a template (or all templates). It's just specific to the current state of the screen.

mehne commented:
As all Threads are closed and redirected here for "Can't add favicon without customizer", and we only get the sulotion by either visit the customize.php manually or use the Site-Icon block 

https://github.com/WordPress/gutenberg/issues/29126
and 
https://core.trac.wordpress.org/ticket/54370

I need to post this here:

1. Manually visit customize.php is not a good solution for non-tech people
2. Only beeing able to set the favicon via the Logo-Block is not viable cause in many cases the Logo is not a square or  in some cases there isn't even a logo added through the FSE

So there should definetly be another solution to add a favicon for non-tech users.

I think the FSE / Gutenberg was added to reduce the custom solutions made by Themes.
But I'm afraid there are gonna be multiple custom solutions even through (unneeded) plugins to add the favicon in a more central and user-friendly way.
